internal/controller/operator/factory/vmdistributedcluster/vmagent.go

```diff
@@ -0,0 +1,298 @@
+	}
+	return nil
+}
+
+func fetchVMAgentDiskBufferMetric(ctx context.Context, httpClient *http.Client, metricsPath string) (int64, error) {
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, metricsPath, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create request for VMAgent at %s: %w", metricsPath, err)
+	}
+	resp, err := httpClient.Do(request)
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch metrics from VMAgent at %s: %w", metricsPath, err)
+	}
+	defer resp.Body.Close()
+	metricBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read VMAgent metrics at %s: %w", metricsPath, err)
+	}
+	metrics := string(metricBytes)
+	for _, metric := range strings.Split(metrics, "\n") {
+		value, found := strings.CutPrefix(metric, VMAgentQueueMetricName)
+		if found {
+			value = strings.Trim(value, " ")
+			res, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return 0, fmt.Errorf("could not parse metric value %s: %w", metric, err)
+			}
+			return int64(res), nil
+		}
+	}
+	return 0, fmt.Errorf("metric %s not found", VMAgentQueueMetricName)
+}
+
+// updateOrCreateVMAgent ensures that the VMAgent is updated or created based on the provided VMDistributedCluster.
+func updateOrCreateVMAgent(ctx context.Context, rclient client.Client, cr *vmv1alpha1.VMDistributedCluster, scheme *runtime.Scheme, vmClusters []*vmv1beta1.VMCluster) (*vmv1beta1.VMAgent, error) {
+	// Get existing vmagent obj using Name and cr namespace
+	vmAgentExists := true
+	vmAgentNeedsUpdate := false
+	vmAgentObj := &vmv1beta1.VMAgent{}
+	namespacedName := types.NamespacedName{Name: cr.Spec.VMAgent.Name, Namespace: cr.Namespace}
+	if err := rclient.Get(ctx, namespacedName, vmAgentObj); err != nil {
+		if k8serrors.IsNotFound(err) {
+			vmAgentExists = false
+			// If it doesn't exist, initialize object for creation.
+			vmAgentObj = &vmv1beta1.VMAgent{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      cr.Spec.VMAgent.Name,
+					Namespace: cr.Namespace,
+				},
+			}
+		} else {
+			return nil, fmt.Errorf("failed to get VMAgent %s/%s: %w", cr.Namespace, cr.Spec.VMAgent.Name, err)
+		}
+	}
+
+	// Prepare the desired spec for VMAgent.
+	desiredVMAgentSpec := vmv1alpha1.CustomVMAgentSpec{}
+	if cr.Spec.VMAgent.Spec != nil {
+		desiredVMAgentSpec = *cr.Spec.VMAgent.Spec.DeepCopy()
+	}
+
+	// Determine tenant id (default to "0" when not provided)
+	tenantID := "0"
+	if cr.Spec.VMAgent.TenantID != nil && *cr.Spec.VMAgent.TenantID != "" {
+		tenantID = *cr.Spec.VMAgent.TenantID
+	}
+
+	tenantPtr := &tenantID
+
+	// Point VMAgent to all VMClusters by constructing RemoteWrite entries
+	if len(vmClusters) > 0 {
+		desiredVMAgentSpec.RemoteWrite = make([]vmv1alpha1.CustomVMAgentRemoteWriteSpec, len(vmClusters))
+		for i, vmCluster := range vmClusters {
+			desiredVMAgentSpec.RemoteWrite[i].URL = remoteWriteURL(vmCluster, tenantPtr)
+		}
+	}
+
+	// Assemble new VMAgentSpec
+	newVMAgentSpec := vmv1beta1.VMAgentSpec{}
```

Comment:
what do you think about marshalling src and unmarshalling to target?

Reply:
That would fit overrides, but not very suitable for new object declaration